### ISSUE TITLE
depends_on :osxfuse for macFUSE

### ISSUE THREAD

**Issue by hkob** [Homebrew/brew#9401]
# Feature suggestion

FUSE for macOS (called as osxfuse) is now [macFUSE](https://osxfuse.github.io/).
Many formulas were depended on osxfuse, so we wrote "depends_on :osxfuse" in each formula.

However, if FUSE for macOS was updated in the System Preferences, we can not install their formulas.
```
sshfs: FUSE for macOS is required for this software. OsxfuseRequirement unsatisfied!
You can install the necessary cask with:
  brew install --cask osxfuse
```

I want to pass the requirement check of macFUSE as well as FUSE for macOS(osxfuse).

# A detailed description of the proposed feature

The requirement check is executed in Library/Homebrew/extend/os/mac/requirements/osxfuse_requirement.rb.
In macFUSE, `fuse.h` is moved from `/usr/local/include/osxfuse` to `/usr/local/include/fuse`.
[I patched the file](https://github.com/hkob/brew/commit/424f06767fda979cd414a06bd2baad6bc17e87d7) in my private repository.
As a result, `sshfs`, which use the macFUSE, can be installed.

# The motivation for the feature

`macFUSE 4` supports Apple Silicon natively.
We can install several formulas depends on FUSE, as Apple Silicon native applications.

# How the feature would be relevant to at least 90% of Homebrew users

Unfortunately, I don't think the M1 Mac has reached 90% yet.

# What alternatives to the feature have been considered

Nothing in particular.

**Comment by hkob** [comment]
Additional Information

Since fuse_darwin.h was removed from macFUSE, I compiled sshfs with [a patch that removes the corresponding includes](https://github.com/hkob/homebrew-m1-beta-brew/blob/main/m1-beta-sshfs.rb).

**Comment by MikeMcQuaid** [comment]
We are removing support for OSXFuse because it is no longer open source software.

**Comment by sunknudsen** [comment]
@MikeMcQuaid Are you planning on deprecating `depends_on :osxfuse` which currently asks users to install FUSE for macOS using `brew install --cask osxfuse`?

**Comment by MikeMcQuaid** [comment]
@sunknudsen Yes.

**Comment by sunknudsen** [comment]
@MikeMcQuaid When are you planning on removing `depends_on :osxfuse` from the codebase?

Although I understand (and value) the motive, this is pretty bad from a UX perspective... as a workaround, would it be possible to implement `depends_on cask:` (example: `depends_on cask: "osxfuse"`) in the context of formulas? An equivalent exists for casks.

We just published a tap for Borg users (see [#5522](https://github.com/borgbackup/borg/issues/5522)) to address the deprecation of FUSE for macOS... the fact running `brew install borgbackup/tap/borgbackup-llfuse` gracefully handles the `osxfuse` cask dependency (thanks to `depends_on :osxfuse`) is very useful.

**Comment by MikeMcQuaid** [comment]
> @MikeMcQuaid When are you planning on removing `depends_on :osxfuse` from the codebase?

I've opened a PR to deprecate it now: https://github.com/Homebrew/brew/pull/9403. We'll remove it from Homebrew/homebrew-core before we deprecate it, though.

> Although I understand (and value) the motive, this is pretty bad from a UX perspective

We cannot have things that have non-open-source dependencies in Homebrew/homebrew-core. We're also not willing to install that stuff in our CI for the same reasons. Given that: it's worse from a UX perspective to have software that is not tested in Homebrew/homebrew-core.

> would it be possible to implement `depends_on cask:` (example: `depends_on cask: "osxfuse"`) in the context of formulas?

Not in Homebrew/homebrew-core. We're intentionally removing all support for formulae that need casks to function.

> We just published a tap for Borg users (see [#5522](https://github.com/borgbackup/borg/issues/5522)) to address the deprecation of FUSE for macOS... the fact running `brew install borgbackup/tap/borgbackup-llfuse` gracefully handles the `osxfuse` cask dependency (thanks to `depends_on :osxfuse`) is very useful.

You can copy the existing `Requirement` logic (complying with license terms) into your tap and it will continue to work as expected with `depends_on BorgOsxfuseRequirement` or similar.

**Comment by sunknudsen** [comment]
@MikeMcQuaid Thanks for the follow-up.

> You can copy the existing Requirement logic (complying with license terms) into your tap and it will continue to work as expected with depends_on BorgOsxfuseRequirement or similar.

Can you please ELI5 the above? These upstream changes are affecting thousands of users... have you published some documentation (intended for developers with limited Homebrew experience) on how to handle things gracefully?

**Comment by MikeMcQuaid** [comment]
> Can you please ELI5 the above?

I don't know what this means.

> These upstream changes are affecting thousands of users... have you published some documentation (intended for developers with limited Homebrew experience) on how to handle things gracefully?

No, we haven't.

**Comment by sunknudsen** [comment]
@MikeMcQuaid ELI5 👉 Explain Like I'm Five.

I am trying my best to keep up with these upstream changes, but feel overwhelmed... I absolutely love Homebrew. The project has made installing open source dependencies so much easier. That being said, I have spent over 4 hours trying to handle this deprecation gracefully leaning about creating formulas and taps (as a Homebrew user vs developer, I knew nothing about that side of the story). I am having a hard time understanding how to handle this change and feel like others would benefit from an official recommendation (with some boilerplate).

To be specific, I don't know where to start to implement the following.

> You can copy the existing Requirement logic (complying with license terms) into your tap and it will continue to work as expected with depends_on BorgOsxfuseRequirement or similar.

**Comment by MikeMcQuaid** [comment]
> @MikeMcQuaid ELI5 👉 Explain Like I'm Five.

Sorry, I'm volunteering on Homebrew on top of my full-time job and cannot justify the time in doing so.

> I am having a hard time understanding how to handle this change and feel like others would benefit from an official recommendation (with some boilerplate).

Sorry, I'm not able to provide more details than I have above. You'll need to figure it out for yourself. I'm happy to answer more specific questions if you look at the code. 

https://docs.brew.sh/Building-Against-Non-Homebrew-Dependencies might help you. https://docs.brew.sh/ and https://rubydoc.brew.sh may do. Beyond that, I cannot help, sorry.

**Comment by MikeMcQuaid** [comment]
> but feel overwhelmed... I absolutely love Homebrew. The project has made installing open source dependencies so much easier.

Thanks for the kind words here, though ❤️

**Comment by EricFromCanada** [comment]
You can essentially copy and paste the contents of the `OsxfuseRequirement` class from its source files into the top of the [formula file](https://github.com/borgbackup/homebrew-tap/blob/master/Formula/borgbackup-llfuse.rb) itself, and then change `depends_on osxfuse: :build` to `depends_on BorgOsxfuseRequirement => :build` (or whatever you rename the class to). See [these](https://discourse.brew.sh/t/how-to-make-formula-depend-on-cask/3557) [threads](https://discourse.brew.sh/t/checking-for-external-existance-of-dependency/6629) for examples of [custom](https://github.com/jonchang/homebrew-biology/blob/231dd3df2d3894c2a9ccab1b087d60af7a4ddde8/Formula/sowhat.rb#L1-L12) [requirements](https://github.com/gfguthrie/homebrew-stdenv/blob/master/Formula/docfx.rb). Further exploration is best done in [Homebrew's Discussions](https://github.com/Homebrew/discussions/discussions).

**Comment by sunknudsen** [comment]
Thanks for helping out @EricFromCanada. 🤓

**Comment by sunknudsen** [comment]
Thanks again for your help @MikeMcQuaid and @EricFromCanada. I was able to put together a formula thanks to your guidance.

To anyone looking for some boilerplate, check-out https://github.com/borgbackup/homebrew-tap/blob/master/Formula/borgbackup-fuse.rb.
